compiler/optimizer/OMRCFGSimplifier.cpp

```diff
@@ -333,6 +333,198 @@ bool OMR::CFGSimplifier::simplifyInstanceOfTestToCheckcast(bool needToDuplicateT
    return true;
    }
 
+bool OMR::CFGSimplifier::simplifyBoundCheckWithThrowException(bool needToDuplicateTree) 
+   {
+   if (trace())
+      traceMsg(comp(), "Start simplifyBoundCheckWithThrowException\n");
+   TR::TreeTop * treeTop = getLastRealTreetop(_block);
+   if (!treeTop)
+      return false;
+   TR::Node *compareNode = treeTop->getNode();
+
+   TR::ILOpCodes opCode = compareNode->getOpCodeValue();
+   if (opCode != TR::ificmplt && opCode != TR::ificmpge)
+      return false;
+
+   bool isCompareLT = opCode == TR::ificmplt;
+   bool patternMatched = false;
+
+   TR::Block *throwBlock = NULL;
+   TR::Block *nextCompareBlock = NULL;
+   TR::Node *firstChildBndChk = NULL;
+   TR::Node *secondChildBndChk = NULL;
+
+   if (compareNode->getNumChildren() > 1) {
+      TR::Node *firstChild = compareNode->getFirstChild();
+      TR::Node *secondChild = compareNode->getSecondChild();
+
+      if (firstChild->getOpCodeValue() == TR::iload) {
+         if (secondChild->getOpCodeValue() == TR::iconst && secondChild->getConstValue() == 0) {
+            patternMatched = true;
+            throwBlock = isCompareLT ? compareNode->getBranchDestination()->getNode()->getBlock() : _block->getNextBlock();
+            nextCompareBlock = isCompareLT ? _block->getNextBlock() : compareNode->getBranchDestination()->getNode()->getBlock();
+            firstChildBndChk = compareNode->getFirstChild();
+         } else if (secondChild->getOpCode().isIntegralLoadVar() || secondChild->getOpCode().isLoadIndirect()) {
+            patternMatched = true;
+            throwBlock = isCompareLT ? _block->getNextBlock() : compareNode->getBranchDestination()->getNode()->getBlock();
+            nextCompareBlock = isCompareLT ? compareNode->getBranchDestination()->getNode()->getBlock() : _block->getNextBlock();
+            secondChildBndChk = compareNode->getSecondChild();
+         }
+      }
+   }
+
+   if (!patternMatched)
+      return false;
+
+   if (trace())
+      traceMsg(comp(), "   Matched with first %s in block_%d\n", isCompareLT ? "ificmplt" : "ificmpge",
+                  _block->getNumber());
+
+   TR::Node *secondCompareNode = nextCompareBlock->getLastRealTreeTop()->getNode();
+
+   if (secondCompareNode->getOpCodeValue() != TR::ificmplt && secondCompareNode->getOpCodeValue() != TR::ificmpge);
+      return false;
```

Comment:
There needs to be a trace statement in this vicinity

Reply:
Added a trace.